File: CP website/Leetcode/14.longest_common_prefix.py
# https://leetcode.com/problems/longest-common-prefix
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        
        ans = []
        for i in zip(*strs):
            
            if len(set(i))==1:
                ans.append(i[0])
            else:break
                
        return "".join(ans)
    

#         di = {}
#         ans = ''
#         n = len(strs)
        
#         for i in range(len(strs)):
#             string = strs[i]
#             for j in range(len(string)):
                
#                 if string[:j+1] in di:
#                     di[string[:j+1]] += 1
#                 else:
#                     di[string[:j+1]] = 1
                
#                 if di[string[:j+1]] == n and len(string[:j+1])>len(ans):
#                     ans = string[:j+1]         
                
#         return ans
                
                

File: CP website/Leetcode/14. Longest Common Prefix.java
// https://leetcode.com/problems/longest-common-prefix/
class Solution {
    public String longestCommonPrefix(String[] strs) {
        Map<String,Integer> map = new HashMap<>();
        for (int i=0;i<strs.length;i++){
            String s = strs[i];
            int n = s.length();
            for(int j=0;j<s.length();j++){
                String sub = s.substring(0,j+1);
                if(map.containsKey(sub)){
                    map.put(sub,map.get(sub)+1);
                }
                else{
                    map.put(sub,1);
                }
            }
        }
                        System.out.println(map);
        int lenk = 0;
        int val = 0;
        String newKey = "";
        for(String key : map.keySet()){
            if(map.get(key) > val){
                val = map.get(key);
                newKey = key;
            }
            if(map.get(key) == val && key.length() > newKey.length()){
                newKey = key;
            }
        }
        if(map.getOrDefault(newKey,0) == strs.length){
            return newKey;
        }
        return "";
        
    }
}class Solution {
    public String longestCommonPrefix(String[] strs) {
        Map<String,Integer> map = new HashMap<>();
        for (int i=0;i<strs.length;i++){
            String s = strs[i];
            int n = s.length();
            for(int j=0;j<s.length();j++){
                String sub = s.substring(0,j+1);
                if(map.containsKey(sub)){
                    map.put(sub,map.get(sub)+1);
                }
                else{
                    map.put(sub,1);
                }
            }
        }
                        System.out.println(map);
        int lenk = 0;
        int val = 0;
        String newKey = "";
        for(String key : map.keySet()){
            if(map.get(key) > val){
                val = map.get(key);
                newKey = key;
            }
            if(map.get(key) == val && key.length() > newKey.length()){
                newKey = key;
            }
        }
        if(map.getOrDefault(newKey,0) == strs.length){
            return newKey;
        }
        return "";
        
    }
}